### PR TITLE
Move goto_programt::(const_)targett comparison to a struct

### PR DESCRIPTION
**Pull request by tautschnig**
This makes sure no one can inadvertently use less-than comparison on
those iterators, which may lead to non-reproducible behaviour
(see #6782).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/goto_program.h [hunk 1]
```diff
@@ -399,6 +399,33 @@ class goto_programt
     typedef std::list<targett> targetst;
     typedef std::list<const_targett> const_targetst;
 
+    /// A total order over `targett` and `const_targett`. Note that the specific
+    /// ordering may wary from execution to execution for it uses comparison on
```
Comment (SaswatPadhi) [inline src/goto-programs/goto_program.h]: minor typo

```suggestion
    /// ordering may vary from execution to execution for it uses comparison on
```

**Review by SaswatPadhi — APPROVED** [review]
Thanks for fixing this throughout, Michael!

File: src/goto-programs/goto_program.h (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/goto_program.h]: Thank you!

**Review by feliperodri — APPROVED** [review]
Thanks, @tautschnig! Changes in `contracts` module LGTM.

**Review by peterschrammel — APPROVED** [review]

**Review by TGWDB — APPROVED** [review]

**Pull request merged**